### PR TITLE
Analyzing leaf partitions of multi-level partition table causes resampling of intermediate partition.

### PR DESCRIPTION
**Pull request by chandankunal**
Problem: Analyzing leaf partitions of multi-level partition table causes
resampling of intermediate partition.

Solution: Remove the intermediate partition from the list of table to be analyzed.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by chandankunal** [comment]
I have opened it as Draft PR as i see some failures in pipeline. Evaluating as this approach is OK or should i make a list of all intermediate table in expand_vacuum_rel() and then keep passing this list till do_analyze_rel().

**Comment by chrishajas** [comment]
I don't think this is the right approach. We want to prevent unnecessary re-sampling of the intermediate partition, but with this PR we'll never sample _any_ partitioned table. For example, if some leaf stats don't have statistics and a user specifies  'ANALYZE ROOTPARTITION <root>`, we do want to sample the root partition.

You may want to take a look at `expand_vacuum_rel()`. Around L1030, we expand the relation to also update the parent partition's stats, but also get the intermediate partition. Is that correct? How does that compare with the behavior in 6X? This code block was changed from 6X, which is ok, but the end user behavior is different and incorrect which we want to fix.

File: src/backend/commands/vacuum.c [hunk 1]
```diff
@@ -1045,11 +1045,16 @@ expand_vacuum_rel(VacuumRelation *vrel, int options)
 				if (!leaf_parts_analyzed(parent_relid, child_relid, vrel->va_cols, elevel))
 					break;
 
-				oldcontext = MemoryContextSwitchTo(vac_context);
-				vacrels = lappend(vacrels, makeVacuumRelation(vrel->relation,
-															  parent_relid,
-															  vrel->va_cols));
-				MemoryContextSwitchTo(oldcontext);
+				if(!(get_rel_relkind(parent_relid) == RELKIND_PARTITIONED_TABLE &&
```
Comment (chrishajas) [inline src/backend/commands/vacuum.c:1048]: The logic looks correct here, but can we refactor this a bit? We're now making 2 calls to `get_rel_relispartition(parent_relid);`. Also add a comment

Comment (chrishajas) [reply src/backend/commands/vacuum.c:1048]: Is `get_rel_relkind(parent_relid) == RELKIND_PARTITIONED_TABLE` ever false here?

Comment (chandankunal) [reply src/backend/commands/vacuum.c:1048]: > Is `get_rel_relkind(parent_relid) == RELKIND_PARTITIONED_TABLE` ever false here?

No it cann't be in this loop as it starts with one level up from leaf. I will remove it.

**Comment by chandankunal** [comment]
Pipeline is green: https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/gpdb-dev-chandankunal-rhel8

I went through src/test/regress/sql/analyze.sql and src/test/regress/sql/incremental_analyze.sql and found it has all relevant test scenarios of this PR. So did not updated it. Although if it is required then i think following unit test case can be added.

DDL:
 create table multipart(a int)  partition by range(a);
 create table part2(a int) partition by range(a);
 create table part3(a int) partition by range(a);
 create table p1(a int);
 create table p2(a int);
 create table p3(a int);
 create table p4(a int);
 alter table multipart attach partition part2 for values from (0) to (10);
 alter table multipart attach partition part3 for values from (10) to (20);
 alter table part2 attach partition p1 for values from (0) to (5);
 alter table part2 attach partition p2 for values from (5) to (10);
 alter table part3 attach partition p3 for values from (10) to (15);
 alter table part3 attach partition p4 for values from (15) to (20);
 insert into multipart select i%20 from generate_series(1,10000)i;

analyze verbose p1;
INFO:  analyzing "public.p1"
INFO:  Executing SQL: select pg_catalog.gp_acquire_sample_rows(17317, 10000, 'f');
ANALYZE

chandank1=# analyze verbose p2;
INFO:  analyzing "public.p2"
INFO:  Executing SQL: select pg_catalog.gp_acquire_sample_rows(17320, 10000, 'f');
INFO:  analyzing "public.multipart" inheritance tree
ANALYZE

chandank1=# analyze verbose p3;
INFO:  analyzing "public.p3"
INFO:  Executing SQL: select pg_catalog.gp_acquire_sample_rows(17323, 10000, 'f');
INFO:  analyzing "public.multipart" inheritance tree
ANALYZE

chandank1=# analyze verbose p4;
INFO:  analyzing "public.p4"
INFO:  Executing SQL: select pg_catalog.gp_acquire_sample_rows(17326, 10000, 'f');
INFO:  analyzing "public.multipart" inheritance tree
ANALYZE

chandank1=# analyze verbose part2;
WARNING:  skipping "part2" --- cannot analyze a mid-level partition. Please run ANALYZE on the root partition table.
ANALYZE

chandank1=# analyze verbose multipart;
INFO:  analyzing "public.p4"
INFO:  Executing SQL: select pg_catalog.gp_acquire_sample_rows(17326, 10000, 'f');
INFO:  analyzing "public.p3"
INFO:  Executing SQL: select pg_catalog.gp_acquire_sample_rows(17323, 10000, 'f');
INFO:  analyzing "public.p2"
INFO:  Executing SQL: select pg_catalog.gp_acquire_sample_rows(17320, 10000, 'f');
INFO:  analyzing "public.p1"
INFO:  Executing SQL: select pg_catalog.gp_acquire_sample_rows(17317, 10000, 'f');
INFO:  analyzing "public.multipart" inheritance tree
ANALYZE

chandank1=# analyze verbose rootpartition multipart;
INFO:  analyzing "public.multipart" inheritance tree
ANALYZE

chandank1=# set optimizer_analyze_midlevel_partition=on;
SET

chandank1=# analyze verbose p1;
INFO:  analyzing "public.p1"
INFO:  Executing SQL: select pg_catalog.gp_acquire_sample_rows(17317, 10000, 'f');
INFO:  analyzing "public.part2" inheritance tree
INFO:  Executing SQL: select pg_catalog.gp_acquire_sample_rows(17311, 10000, 't');
INFO:  analyzing "public.multipart" inheritance tree
ANALYZE

chandank1=# analyze verbose p2;
INFO:  analyzing "public.p2"
INFO:  Executing SQL: select pg_catalog.gp_acquire_sample_rows(17320, 10000, 'f');
INFO:  analyzing "public.part2" inheritance tree
INFO:  Executing SQL: select pg_catalog.gp_acquire_sample_rows(17311, 10000, 't');
INFO:  analyzing "public.multipart" inheritance tree
ANALYZE

chandank1=# analyze verbose p3;
INFO:  analyzing "public.p3"
INFO:  Executing SQL: select pg_catalog.gp_acquire_sample_rows(17323, 10000, 'f');
INFO:  analyzing "public.part3" inheritance tree
INFO:  Executing SQL: select pg_catalog.gp_acquire_sample_rows(17314, 10000, 't');
INFO:  analyzing "public.multipart" inheritance tree
ANALYZE

chandank1=# analyze verbose p4;
INFO:  analyzing "public.p4"
INFO:  Executing SQL: select pg_catalog.gp_acquire_sample_rows(17326, 10000, 'f');
INFO:  analyzing "public.part3" inheritance tree
INFO:  Executing SQL: select pg_catalog.gp_acquire_sample_rows(17314, 10000, 't');
INFO:  analyzing "public.multipart" inheritance tree
ANALYZE

chandank1=# analyze verbose part2;
INFO:  analyzing "public.p2"
INFO:  Executing SQL: select pg_catalog.gp_acquire_sample_rows(17320, 10000, 'f');
INFO:  analyzing "public.p1"
INFO:  Executing SQL: select pg_catalog.gp_acquire_sample_rows(17317, 10000, 'f');
INFO:  analyzing "public.part2" inheritance tree
INFO:  Executing SQL: select pg_catalog.gp_acquire_sample_rows(17311, 10000, 't');
INFO:  analyzing "public.multipart" inheritance tree
ANALYZE

chandank1=# analyze verbose part3;
INFO:  analyzing "public.p4"
INFO:  Executing SQL: select pg_catalog.gp_acquire_sample_rows(17326, 10000, 'f');
INFO:  analyzing "public.p3"
INFO:  Executing SQL: select pg_catalog.gp_acquire_sample_rows(17323, 10000, 'f');
INFO:  analyzing "public.part3" inheritance tree
INFO:  Executing SQL: select pg_catalog.gp_acquire_sample_rows(17314, 10000, 't');
INFO:  analyzing "public.multipart" inheritance tree
ANALYZE

chandank1=# analyze verbose multipart;
INFO:  analyzing "public.p4"
INFO:  Executing SQL: select pg_catalog.gp_acquire_sample_rows(17326, 10000, 'f');
INFO:  analyzing "public.p3"
INFO:  Executing SQL: select pg_catalog.gp_acquire_sample_rows(17323, 10000, 'f');
INFO:  analyzing "public.p2"
INFO:  Executing SQL: select pg_catalog.gp_acquire_sample_rows(17320, 10000, 'f');
INFO:  analyzing "public.p1"
INFO:  Executing SQL: select pg_catalog.gp_acquire_sample_rows(17317, 10000, 'f');
INFO:  analyzing "public.part3" inheritance tree
INFO:  Executing SQL: select pg_catalog.gp_acquire_sample_rows(17314, 10000, 't');
INFO:  analyzing "public.part2" inheritance tree
INFO:  Executing SQL: select pg_catalog.gp_acquire_sample_rows(17311, 10000, 't');
INFO:  analyzing "public.multipart" inheritance tree
ANALYZE

chandank1=# analyze verbose rootpartition multipart;
INFO:  analyzing "public.multipart" inheritance tree
ANALYZE

File: src/backend/commands/analyze.c [hunk 2]
```diff
@@ -700,6 +700,7 @@ do_analyze_rel(Relation onerel, VacuumParams *params,
 	}
 
 	sample_needed = needs_sample(onerel, vacattrstats, attr_cnt);
+
```
Comment (pobbatihari) [inline src/backend/commands/analyze.c:703]: remove extra line

**Comment by chrishajas** [comment]
> Pipeline is green: https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/gpdb-dev-chandankunal-rhel8
> 
> I went through src/test/regress/sql/analyze.sql and src/test/regress/sql/incremental_analyze.sql and found it has all relevant test scenarios of this PR. So did not updated it. Although if it is required then i think following unit test case can be added.
> 

Does this PR impact whether midlevel stats are collected in multi-level partition tables? For example, before this we'd sample (and collect) mid level stats when the guc is disabled. I think we can check for that in a test (possibly add a check in an existing test?). Also if the mid level guc is enabled, we should see stats populated. If there's a behavior change, we should add a corresponding test case.

File: src/backend/commands/analyze.c (hunk 2 above)
Comment (chandankunal) [reply src/backend/commands/analyze.c]: Done.

**Review by chrishajas — APPROVED** [review]
LGTM after feedback addressed

File: src/backend/commands/vacuum.c [hunk 3]
```diff
@@ -1037,6 +1037,7 @@ expand_vacuum_rel(VacuumRelation *vrel, int options)
 				int			elevel = ((options & VACOPT_VERBOSE) ? LOG : DEBUG2);
 
 				parent_relid = get_partition_parent(child_relid);
+				ispartition = get_rel_relispartition(parent_relid);
```
[the remark this answers is not in the document]
Comment (chandankunal) [reply src/backend/commands/vacuum.c:1040]: I would have changed it if it was do...while loop But actually "while (ispartition)" condition is for the current table in that iteration. Also I wanted to avoid new variable like child_isPartition and parent_isPartition just like parent_relid and child_relid.

**Review by pobbatihari — APPROVED** [review]

File: src/backend/commands/vacuum.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (chrishajas) [reply src/backend/commands/vacuum.c:1040]: I see, then we should add a comment. In its current state it's a bit hard to follow since for L1053 it's being used as "is the parent a partition"

[the remark this answers is not in the document]
Comment (chandankunal) [reply src/backend/commands/vacuum.c:1040]: Done.
